### PR TITLE
Refactor parameter handling for *Request classes.

### PR DESCRIPTION
**Pull request by coryan**
Too much duplication, reduce the amount of code to write.

The interesting bits are in `request_parameters.h`, though to see
how they are used look at the new implementation for `GetBucketMetadataRequest`.

### CODE REVIEW

File: google/cloud/storage/internal/request_parameters.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  void AddToHttpRequest(HttpRequest& request) const {
+    request.AddWellKnownParameter(parameter_);
+    RequestParameterList<Parameters...>::AddToHttpRequest(request);
+  }
+
+ private:
+  Parameter parameter_;
+};
+
+template <typename Parameter>
```
Comment (houglum) [inline google/cloud/storage/internal/request_parameters.h]: Nit: Could you place this class declaration first in the file?  That way, the ordering in this file is more consistent with the inheritance hierarchy, starting at the base.

File: google/cloud/storage/internal/request_parameters.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  void set_parameter(Parameter&& p) { parameter_ = std::move(p); }
+
+  template <typename HttpRequest>
+  void AddToHttpRequest(HttpRequest& request) const {
+    request.AddWellKnownParameter(parameter_);
+    RequestParameterList<Parameters...>::AddToHttpRequest(request);
+  }
+
+ private:
+  Parameter parameter_;
```
Comment (houglum) [inline google/cloud/storage/internal/request_parameters.h]: Q: How does this work? Is there an individual private member (singular, not a parameter pack) for each `Parameter`?  Or is there one a member of type `Parameter`, and the underlying base class's parameter of type `Parameters...`?

File: google/cloud/storage/internal/insert_object_media_request.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+/**
+ * Request the metadata for a bucket.
+ *
+ * TODO(#710) - add missing request parameters.
+ */
+class InsertObjectMediaRequest
+    : private RequestParameters<Generation, IfGenerationMatch,
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/insert_object_media_request.h]: Sorry it took me two tries to figure out the right way.  That will happen a lot btw.  I prefer to try something wrong and fix it later.

File: google/cloud/storage/internal/get_bucket_metadata_request.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+   * @param p the parameter values
+   * @return The object after all the parameters have been changed.
+   */
+  template <typename... Parameters>
+  GetBucketMetadataRequest& set_multiple_parameters(Parameters&&... p) {
+    RequestParameters::set_multiple_parameters(std::forward<Parameters>(p)...);
+    return *this;
+  }
+
+  using RequestParameters::AddToHttpRequest;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/get_bucket_metadata_request.h]: Done.

File: google/cloud/storage/internal/request_parameters.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/request_parameters.h]: Done.  That requires forward declaring the template first, then specializing it, and then defining it.  Let me know what you think.

File: google/cloud/storage/internal/request_parameters.h (hunk 2 above)
Comment (houglum) [reply google/cloud/storage/internal/request_parameters.h]: Ohhhh.  Thanks for the explanation!

File: google/cloud/storage/internal/request_parameters.h [hunk 5]
```diff
@@ -21,37 +21,51 @@
...
   Parameter parameter_;
 };
 
-template <typename Parameter>
-class RequestParameterList<Parameter> {
+/**
+ * Refactor common functions that manipulate list of parameters in a request.
+ *
+ * This class is used in the implementation of `RequestParameters` see below for
+ * more details.
```
Comment (houglum) [inline google/cloud/storage/internal/request_parameters.h]: Thanks for the additional comments.  This is way more readable than the first time I looked over it :)

**Pull request merged**